_posts/2019-04-24-server-rendering-responsively.md

```diff
@@ -0,0 +1,199 @@
+---
+
+At Artsy, we use server-side rendering (SSR) to deliver every page you hit on [artsy.net](https://artsy.net). We
+decided on using SSR for many reasons, amongst them performance. (TODO: Maybe more about why we chose SSR, but I
+don't know much about that history.)
+
+(TODO: Maybe a bit more about SSR & `hydrate`, or just link to those docs.)
+
+One challenge we've run into with SSR is rendering markup and styles from the server that the browser doesn't have
+to throw away. When your browser finishes loading our JavaScript & hydrates our React app, our app suddenly has
+more information about your browser dimensions than we had in the server request. React will go through the
+reconciliation process with this new information, to determine which DOM elements need to be updated. If there are
+significant differences between the markup our server gave you and the component tree our React app is now
+rendering, there's noticeable churn in the DOM.
+
+We'd like our app to churn as little as possible during hydration. We've taken several steps (and mis-steps) toward
+accomplishing this goal: how can we emit _usable_ responsive markup from the server, so that the browser doesn't
+have to throw everything away when the browser dimensions become known?
+
+To help us solve this problem, we've released
+[@artsy/react-responsive-media](https://github.com/artsy/react-responsive-media). This is the story of how
+@artsy/react-responsive-media came about.
+
+<!-- more -->
+
+## Iteration 1: Render props and a disposable component tree
+
+Our initial attempt to emit responsive content from the server is what prompted us to dig deeper into this problem.
+We started with a `<Responsive>` component that used
+[a "render props"/"function as a child" pattern](https://reactjs.org/docs/render-props.html#using-props-other-than-render),
+and allowed you to return different JSX based on the current breakpoint. In practice, it looked something like
+this:
+
+```
+<Responsive>
+  {({ xs }) => {
+    if (xs) {
+      return <Box width={1} />
+    } else {
+      return <Box width={2} />
+    }
+  }}
+</Responsive>
+```
+
+The argument to the child function contained a flag for each of the breakpoints our app supports. In the example
+above, we're rendering a `<Box>` with a width of `1` if the screen matches our `xs` breakpoint; otherwise, we're
+rendering a `<Box>` with a width of `2`.
+
+Our biggest problem with this API was that on the server, we don't know which breakpoint you're at. We had to
+default to something, so we assumed that if we didn't know which breakpoint you wanted, we would give you the
+`xs`/"extra small" one. But if you were browsing with a desktop, the first client-side render will completely throw
+away what the server gave you. In the example above, that's probably not that big a deal, but our `<Responsive>`
+elements tended to contain large sub-trees of components. This is a big deal, and this is the DOM churn we really
+want to avoid.
```

Comment:
While churn is indeed an issue it's important to emphasize that the user might be be getting a totally incorrect layout since the user-agent detection is flimsy, which leads to a perceptible shift in the UI layout on mount.

Reply:
I think there's something more to emphasize here. In react 16 when rehydration is happening it _doesn't_ throw the DOM out. It naively tries to patch it up to match what the client is rendering. This creates really hard to debug issues.

According to the [react docs](https://reactjs.org/docs/react-dom.html#hydrate)

> React expects that the rendered content is identical between the server and the client. It can patch up differences in text content, but you should treat mismatches as bugs and fix them.

So if it was just throwing out markup we would be occurring a bit of a perf penalty but we could've kept going with that solution regardless. The thing is, it _wasn't_ just throwing it out. It was patching the dom in ways that broke the layout. We had really weird, really hard to debug issues like this:

---

![2018-12-11_15-40-34 2018-12-11_15_42_03](https://user-images.githubusercontent.com/3087225/56924726-a14a7f80-6a9b-11e9-969e-5801a8e7cf87.gif)

---

This is a classic rehydration issue where client content (like artist info) starts rendering inside of server rendered containers (spinner) and created really weird visual bugs. So ultimately we couldn't keep going with this solution because it fundamentally doesn't work with rehydration. All the markup _has_ to be the same between server and client. Since the server can't know what breakpoint you're on that means we had to render all the markup for every breakpoint.